### PR TITLE
Lazy parsing

### PR DESCRIPTION
**Pull request by philipc**

### CODE REVIEW

**Comment by main--** [comment]
Ah, I guess the problem is that that would keep a `&mut` borrow on `self` alive?

In that case a (ref)cell may be worth considering - it has the added bonus of not requiring `&mut` for the query functions either.

**Comment by philipc** [comment]
Yeah it was due to the `&mut` borrow, but I just did similar to what we used to do before hardliner, instead of thinking about it much; I'll look into it more tomorrow. `Cell` isn't appropriate because it needs to move the value in and out of the `Cell`. I've not had much experience with `RefCell`, but I think that would be trading `&mut` for runtime panics if you happen to use this from multiple threads, so I don't think that works either?

**Comment by philipc** [comment]
There's also https://crates.io/crates/lazy-init

**Comment by main--** [comment]
>I've not had much experience with `RefCell`, but I think that would be trading `&mut` for runtime panics if you happen to use this from multiple threads, so I don't think that works either?

No, what `RefCell` does is simply that it moves the borrow checker from compile time to runtime (at least that's how I think of it). It's basically the single-threaded version of `RwLock`.

I didn't consider multi-threaded environments. In that case, lazy-init does seem like a better solution.

**Comment by philipc** [comment]
Added a commit for lazy-init.

I looked into using `RefCell` but it was starting to get complicated because the `Ref` needed to live as long as `IterFrames`.

File: src/lib.rs [hunk 1]
```diff
@@ -197,66 +207,76 @@ impl<'a> Context<gimli::EndianBuf<'a, gimli::RunTimeEndian>> {
...
+                self.comp_name.clone(),
+            )?;
+
+            let (lnp, mut sequences) = ilnp.sequences()?;
+            sequences.retain(|x| x.start != 0);
+            sequences.sort_by_key(|x| x.start);
+            Ok(Lines { lnp, sequences })
+        }) {
+            Ok(ref lines) => Ok(lines),
+            Err(e) => Err(e),
```
Comment (main--) [inline src/lib.rs]: Took me a while to figure out what this does. Perhaps `.as_ref().map_err(Error::clone)` is clearer?

Comment (philipc) [reply src/lib.rs]: Yeah that looks better, thanks.

**Pull request merged**